File: HashSet/LongestSubstringWithoutRepeatingCharacters.cpp
#include <iostream>
#include <unordered_set>
#include <algorithm>
using namespace std;

class LongestSubstringWithoutRepeatingCharacters {
public:
    int lengthOfLongestSubstring(const string& s) {
        int n = s.length();
        int maxLength = 0;
        int left = 0;
        unordered_set<char> seenChars;

        for (int right = 0; right < n; ++right) {
            while (seenChars.find(s[right]) != seenChars.end()) {
                seenChars.erase(s[left]);
                ++left;
            }
            seenChars.insert(s[right]);
            maxLength = max(maxLength, right - left + 1);
        }
        return maxLength;
    }
};

int main() {
    longestSubstringWithoutRepeatingCharacters solution;

    string s1 = "abcabcbb";
    string s2 = "bbbbb";

    cout << "Output for s1: " << solution.lengthOfLongestSubstring(s1) << endl; // 3
    cout << "Output for s2: " << solution.lengthOfLongestSubstring(s2) << endl; // 1

    return 0;
}

File: HashSet/LongestSubstringWithoutRepeatingCharacters.java
import java.util.HashSet;

public class LongestSubstringWithoutRepeatingCharacters {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int maxLenght = 0;
        int left = 0;
        HashSet<Character> seenChars = new HashSet<>();

        for(int right = 0; right < n; right++){
            while (seenChars.contains(s.charAt(right))) {
                seenChars.remove(s.charAt(left));
                left++;
            }
            seenChars.add(s.charAt(right));
            maxLenght = Math.max(maxLenght, right - left + 1);
        }
        return maxLenght;
    }

    public static void main(String[] args) {
        LongestSubstringWithoutRepeatingCharacters solution = new LongestSubstringWithoutRepeatingCharacters();

        String s1 = "abcabcbb";
        String s2 = "bbbbb";

        System.out.println("Output for s1: " + solution.lengthOfLongestSubstring(s1)); // 3
        System.out.println("Output for s2: " + solution.lengthOfLongestSubstring(s2)); // 1
    }
}
